parsing_scripts/update.py

```diff
@@ -0,0 +1,66 @@
+import os, json, sys
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+sys.path.append(BASE_DIR)
+os.environ['DJANGO_SETTINGS_MODULE'] = 'MacPorts.settings'
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "MacPorts.settings")
+
+import django
+
+django.setup()
+
+from ports.models import Port, Maintainer, Category, Dependency
+
+
+def open_portindex_json(path='portindex.json'):
+    with open(path, "r") as file:
+        ports = json.load(file)
+    return ports
+
+
+def full_update_ports(ports):
+
+    Maintainer.objects.all().delete()
```

Comment:
Deleting all maintainers sounds too extreme. Maybe we could have some brainstorming together (voice or real-time written chat or something).

Reply:
Yes, even I am looking for other ways of doing this. But updating the maintainers table seems really difficult because we have treated any combination of email and GitHub as a single unique entity. And hence, we cannot find the previous entry and change it.

But there is one more way I could think of:
1. While updating the info of a given port clear all the relations that exist with the maintainers table.
2. Add new maintainers, and new relations.
3. At the end if there happens to be any maintainer in the maintainers table with no ports associated with it -> delete it.

But the same purpose is served by first deleting them all, maybe?

We can do brainstorming as per your availability. Voice and written chat both work for me.

Reply:
I mean ... deleting maintainers themselves seems a bit of an overkill. Also imagine that at some point in the future you'll add additional information like real name, password hash, links to trac tickets etc. Deleting relations between the ports and maintainers is less invasive, but deleting all the maintainers is somewhat problematic. Your approach described above seems better.

I would need to check whether Django supports database transactions. Port update should be a single transaction (if supported; not a big problem if we don't do it as a single transaction for now) which would delete people who are no longer maintainers and add new ones. You only add a new maintainer to the maintainers table if that maintainer doesn't exist already, and then only update the relations between the two.

You could in theory delete all relations at once, but note that if this in fact takes fourty minutes on heroku, the whole webpage will be useless during those fourty minutes.

It would be nice to have an action that takes a list of ports to update. The list could either be ten or thousand ports listed by name, or some kind of signal to say that we should update all ports. I would say that the procedure for updating should be almost the same in both cases, and in any case this is relatively agnostic as to whether it runs on inefficient server or somewhere where all actions are super fast.

Reply:
> I mean ... deleting maintainers themselves seems a bit of an overkill. Also imagine that at some point in the future you'll add additional information like real name, password hash, links to trac tickets etc. Deleting relations between the ports and maintainers is less invasive, but deleting all the maintainers is somewhat problematic. Your approach described above seems better.

Okay, we can also avoid deleting all the relations at once. But the queries are increasing with every bit of complication we add.

> You could in theory delete all relations at once, but note that if this in fact takes fourty minutes on heroku, the whole webpage will be useless during those fourty minutes.

Hmm, that is an important point I missed.

> It would be nice to have an action that takes a list of ports to update. The list could either be ten or thousand ports listed by name, or some kind of signal to say that we should update all ports. I would say that the procedure for updating should be almost the same in both cases, and in any case this is relatively agnostic as to whether it runs on inefficient server or somewhere where all actions are super fast.

Like a POST request with the list of ports? Or just some python command?
If I am getting it right, we take a list of names of ports and pick only those ports from the JSON file- and then we hand over the JSON data to our same `update.py` script?

Reply:
It might be ok to delete the relations between ports and maintainers for one port at once (not the maintainers, just the relations). But it would be super strange to remove all relations at the same time (or worse, all maintainers).

You probably want to implement some subsection of the following (not necessarily all):
* command that updates all ports (and uses the local copy of portindex json)
* command that updates the given list of ports (and uses portindex json)
* command that updates ports from a given json object (and that object contains all data about all ports that you care about updating / adding / removing)

Once you have that functionality in place, there are different alternatives about how you want to run them. Yes, one option would be a post request (you just need to make sure that you don't end up with the whole world being able to spam the applications with requests to update random ports). Another option would be a local script that triggers the update (there could be a local job inspecting git updates, running portindex, portindex2json and making requests for updates). The basic update functionality should not care about how the updates are triggered. It's just the triggers on top that govern this. We could decide to trigger via post actions from the buildbot, or from local macports installation, or something else. As long as it is secure, efficient, reliable, easy to understand, ... I don't care as much how this triggering part is done.

Reply:
> It might be ok to delete the relations between ports and maintainers for one port at once (not the maintainers, just the relations). But it would be super strange to remove all relations at the same time (or worse, all maintainers).

Okay, the latest commit removes only the relations between ports and maintainers for one port at once. All, we will have to do next is take care of the maintainers with zero ports related to them, which is unlikely but if someone tries to get rid of the inconsistency reported by the app- then it is very essential.

> * command that updates all ports (and uses the local copy of portindex json)

We have this.

> * command that updates the given list of ports (and uses portindex json)

Will add this.

> * command that updates ports from a given json object (and that object contains all data about all ports that you care about updating / adding / removing)

We also have this. The functions that update the database just receive a list of JSON objects. So, for full update, we pass it the entire JSON read from local portindex.json, for partial update we pass to it the selected JSON objects and for the third one we pass to it the JSON received.